File: Treachery-iOS/Treachery-iOSTests/RoleTests.swift
import Testing
import Foundation
@testable import Treachery_iOS

struct RoleTests {

    // MARK: - Raw Values & CaseIterable

    @Test func allCasesExist() {
        #expect(Role.allCases.count == 4)
        #expect(Role.allCases.contains(.leader))
        #expect(Role.allCases.contains(.guardian))
        #expect(Role.allCases.contains(.assassin))
        #expect(Role.allCases.contains(.traitor))
    }

    @Test func rawValues() {
        #expect(Role.leader.rawValue == "leader")
        #expect(Role.guardian.rawValue == "guardian")
        #expect(Role.assassin.rawValue == "assassin")
        #expect(Role.traitor.rawValue == "traitor")
    }

    // MARK: - Display Names

    @Test func displayNames() {
        #expect(Role.leader.displayName == "Leader")
        #expect(Role.guardian.displayName == "Guardian")
        #expect(Role.assassin.displayName == "Assassin")
        #expect(Role.traitor.displayName == "Traitor")
    }

    // MARK: - Win Condition Text

    @Test func winConditionTextIsNonEmpty() {
        for role in Role.allCases {
            #expect(!role.winConditionText.isEmpty)
        }
    }

    // MARK: - Identifiable

    @Test func identifiableUsesRawValue() {
        for role in Role.allCases {
            #expect(role.id == role.rawValue)
        }
    }

    // MARK: - Role Distribution

    @Test func distributionForFourPlayers() {
        let dist = Role.distribution(forPlayerCount: 4)
        #expect(dist.leaders == 1)
        #expect(dist.guardians == 0)
        #expect(dist.assassins == 2)
        #expect(dist.traitors == 1)
        #expect(dist.leaders + dist.guardians + dist.assassins + dist.traitors == 4)
    }

    @Test func distributionForFivePlayers() {
        let dist = Role.distribution(forPlayerCount: 5)
        #expect(dist.leaders == 1)
        #expect(dist.guardians == 1)
        #expect(dist.assassins == 2)
        #expect(dist.traitors == 1)
        #expect(dist.leaders + dist.guardians + dist.assassins + dist.traitors == 5)
    }

    @Test func distributionForSixPlayers() {
        let dist = Role.distribution(forPlayerCount: 6)
        #expect(dist.leaders == 1)
        #expect(dist.guardians == 1)
        #expect(dist.assassins == 3)
        #expect(dist.traitors == 1)
        #expect(dist.leaders + dist.guardians + dist.assassins + dist.traitors == 6)
    }

    @Test func distributionForSevenPlayers() {
        let dist = Role.distribution(forPlayerCount: 7)
        #expect(dist.leaders == 1)
        #expect(dist.guardians == 2)
        #expect(dist.assassins == 3)
        #expect(dist.traitors == 1)
        #expect(dist.leaders + dist.guardians + dist.assassins + dist.traitors == 7)
    }

    @Test func distributionForEightPlayers() {
        let dist = Role.distribution(forPlayerCount: 8)
        #expect(dist.leaders == 1)
        #expect(dist.guardians == 2)
        #expect(dist.assassins == 3)
        #expect(dist.traitors == 2)
        #expect(dist.leaders + dist.guardians + dist.assassins + dist.traitors == 8)
    }

    @Test func distributionAlwaysHasOneLeader() {
        for count in 4...8 {
            let dist = Role.distribution(forPlayerCount: count)
            #expect(dist.leaders == 1)
        }
    }

    @Test func distributionDefaultFallback() {
        let dist = Role.distribution(forPlayerCount: 99)
        #expect(dist.leaders == 1)
        #expect(dist.guardians == 0)
        #expect(dist.assassins == 2)
        #expect(dist.traitors == 1)
    }

    // MARK: - Minimum Player Count

    @Test func minimumPlayerCountIsFour() {
        #expect(Role.minimumPlayerCount == 4)
    }

    // MARK: - Codable

    @Test func decodesFromJSON() throws {
        let json = Data(#""leader""#.utf8)
        let role = try JSONDecoder().decode(Role.self, from: json)
        #expect(role == .leader)
    }

    @Test func encodesToJSON() throws {
        let data = try JSONEncoder().encode(Role.assassin)
        let string = String(data: data, encoding: .utf8)
        #expect(string == #""assassin""#)
    }
}

File: TreacheryAndroid/app/src/test/java/com/solomon/treachery/model/RoleTests.kt
package com.solomon.treachery.model

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class RoleTests {

    @Test
    fun `all cases exist`() {
        assertEquals(4, Role.entries.size)
    }

    @Test
    fun `raw values match expected strings`() {
        assertEquals("leader", Role.LEADER.value)
        assertEquals("guardian", Role.GUARDIAN.value)
        assertEquals("assassin", Role.ASSASSIN.value)
        assertEquals("traitor", Role.TRAITOR.value)
    }

    @Test
    fun `display names are human readable`() {
        assertEquals("Leader", Role.LEADER.displayName)
        assertEquals("Guardian", Role.GUARDIAN.displayName)
        assertEquals("Assassin", Role.ASSASSIN.displayName)
        assertEquals("Traitor", Role.TRAITOR.displayName)
    }

    @Test
    fun `win condition text is non-empty for all roles`() {
        Role.entries.forEach { role ->
            assertTrue(role.winConditionText.isNotEmpty(), "${role.name} should have win condition text")
        }
    }

    @Test
    fun `fromValue returns correct role`() {
        assertEquals(Role.LEADER, Role.fromValue("leader"))
        assertEquals(Role.ASSASSIN, Role.fromValue("assassin"))
    }

    @Test
    fun `fromValue returns null for unknown value`() {
        assertNull(Role.fromValue("unknown"))
        assertNull(Role.fromValue(""))
    }

    @Test
    fun `minimum player count is 4`() {
        assertEquals(4, Role.MINIMUM_PLAYER_COUNT)
    }

    // ── Role Distribution ──

    @Test
    fun `distribution for 4 players`() {
        val dist = Role.distribution(4)
        assertEquals(1, dist.leaders)
        assertEquals(0, dist.guardians)
        assertEquals(2, dist.assassins)
        assertEquals(1, dist.traitors)
    }

    @Test
    fun `distribution for 5 players`() {
        val dist = Role.distribution(5)
        assertEquals(1, dist.leaders)
        assertEquals(1, dist.guardians)
        assertEquals(2, dist.assassins)
        assertEquals(1, dist.traitors)
    }

    @Test
    fun `distribution for 6 players`() {
        val dist = Role.distribution(6)
        assertEquals(1, dist.leaders)
        assertEquals(1, dist.guardians)
        assertEquals(3, dist.assassins)
        assertEquals(1, dist.traitors)
    }

    @Test
    fun `distribution for 7 players`() {
        val dist = Role.distribution(7)
        assertEquals(1, dist.leaders)
        assertEquals(2, dist.guardians)
        assertEquals(3, dist.assassins)
        assertEquals(1, dist.traitors)
    }

    @Test
    fun `distribution for 8 players`() {
        val dist = Role.distribution(8)
        assertEquals(1, dist.leaders)
        assertEquals(2, dist.guardians)
        assertEquals(3, dist.assassins)
        assertEquals(2, dist.traitors)
    }

    @Test
    fun `distribution total equals player count for all valid counts`() {
        for (count in 4..8) {
            val dist = Role.distribution(count)
            val total = dist.leaders + dist.guardians + dist.assassins + dist.traitors
            assertEquals(count, total, "Distribution for $count players should total $count")
        }
    }

    @Test
    fun `distribution for out-of-range defaults to 4-player`() {
        val dist = Role.distribution(3)
        val default4 = Role.distribution(4)
        assertEquals(default4, dist)
    }

    @Test
    fun `each role has a distinct color`() {
        val colors = Role.entries.map { it.roleColor }
        assertEquals(colors.size, colors.toSet().size, "All role colors should be unique")
    }
}
